File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/dev/ParallelSpaceHook.kt
package app.getvela.wallet.dev

import androidx.compose.runtime.Composable
import app.getvela.wallet.feature.send.core.UserOpSigner

/**
 * The parallel space's seam into the main source set (spec 043, research D2).
 *
 * The parallel space is the real app with one substitution — where a passkey
 * would sign, the core's fixed keyset signs — behind a door that exists only
 * in DEBUG builds. Main code never names the debug classes: it asks this
 * object, and the build type decides who answers. `ParallelSpaceBinding`
 * (one file under `src/debug`, one under `src/release`) installs a provider
 * at application start; the release one installs nothing, so every method
 * here answers "not active" and the badge draws nothing.
 *
 * A release APK therefore carries neither the door nor a path to the fixture
 * signer; the fixture cdylib and its bindings are debug-only source-set
 * artefacts and are not packaged (FR-001, SC-010).
 */
object ParallelSpaceHook {
    @Volatile
    private var provider: ParallelSpaceProvider? = null

    /** Installed once by the build type's `ParallelSpaceBinding`. */
    fun install(provider: ParallelSpaceProvider?) {
        this.provider = provider
    }

    /** Inside the space right now — the badge answers to this, not to the build type. */
    fun active(): Boolean = provider?.active() == true

    /** The fixture signer when active, else `null` (the real passkey signs). */
    fun signer(): UserOpSigner? = provider?.takeIf { it.active() }?.signer()

    /** The fixture account the space signs in as, when active. */
    fun fixtureAccount(): FixtureAccount? = provider?.takeIf { it.active() }?.fixtureAccount()

    /** Drawn over every route while active; nothing otherwise. */
    @Composable
    fun Badge() {
        provider?.takeIf { it.active() }?.Badge()
    }
}

/** What the debug source set provides; absent in release. */
interface ParallelSpaceProvider {
    fun active(): Boolean

    fun signer(): UserOpSigner

    fun fixtureAccount(): FixtureAccount

    @Composable
    fun Badge()
}

/**
 * One account of the core's fixed keyset — the same scalars, credential ids
 * and derived Safe the web's `/parallel` and the desktop's
 * `VELA_PARALLEL_SPACE=1` use, so a wallet one client created in its space is
 * the wallet another signs for in its own.
 */
data class FixtureAccount(
    val index: Int,
    val credentialIdHex: String,
    val publicKeyHex: String,
    val address: String,
)

File: app-ios/VelaWallet/VelaWallet/Core/ParallelSpaceHook.swift
//
//  ParallelSpaceHook.swift
//  VelaWallet
//
//  The seam between the app and a test environment that does not ship.
//
//  The parallel space is the **real app with one substitution**: where a
//  passkey would sign, `vela-core`'s fixed keyset signs. Same chains, same
//  relay, same storage, same screens. It exists because every feature from spec
//  052 onward has to be verified on the founder's own iPhone, and a send ends
//  in a biometric prompt — which needs a finger, and a finger cannot be
//  scripted.
//
//  ## Why this file is always compiled and the implementation is not
//
//  A Release build must contain neither the keys nor the door (FR-001). That is
//  enforced by the build configuration: `Dev/ParallelSpaceBinding.swift` and
//  the generated `Dev/vela_dev_fixtures.swift` are in
//  `EXCLUDED_SOURCE_FILE_NAMES` for Release, and the fixtures archive is linked
//  only by Debug. So the Release binary cannot reference a fixture symbol —
//  **it would not compile if anything outside this seam tried**, which is a
//  stronger guarantee than a runtime flag.
//
//  This protocol, and the `nil` provider that is all Release ever sees, are the
//  shape that makes that possible.
//

import SwiftUI

@MainActor
protocol ParallelSpaceProvider {
    /// Whether the app is inside the space right now.
    var isActive: Bool { get }

    /// Read the launch environment and the persisted flag, and enter or leave.
    ///
    /// Called **before the session machine boots**, so the record is on disk by
    /// the time anything reads the account list.
    func applyIfRequested(store: VelaStore, accounts: AccountStore) async

    /// The signer to use while the space is active.
    func signer() -> UserOpSigner?
}

enum ParallelSpaceHook {
    /// Installed by the Debug binding at launch. `nil` in every Release build,
    /// and in any Debug build that has not opted in.
    @MainActor static var provider: ParallelSpaceProvider?

    @MainActor static var isActive: Bool { provider?.isActive ?? false }

    @MainActor
    static func applyIfRequested(store: VelaStore, accounts: AccountStore) async {
        await provider?.applyIfRequested(store: store, accounts: accounts)
    }

    /// The signer for a user operation: the fixture's while the space is
    /// active, otherwise the person's own passkey.
    ///
    /// Written as a fallback rather than a branch inside the send path so there
    /// is exactly ONE signing call site, and the substitution is the only
    /// difference between the two environments (FR-002).
    @MainActor
    static func signer(passkey: PasskeyExecutor) -> UserOpSigner {
        provider?.signer() ?? PasskeyUserOpSigner(passkey: passkey)
    }
}

/// The marker that answers to neither gate.
///
/// Deliberately **not a corpus string**: it is a developer marker, it must be
/// conspicuous rather than localised, and it never ships. The desktop and
/// Android made the same call.
///
/// `#if DEBUG` around the view rather than only around its caller, so the
/// literal itself is absent from a Release binary. A dormant "PARALLEL SPACE"
/// string in a shipped wallet is not key material and not a door — but it is
/// the first thing a reviewer would ask about, and the answer should be that
/// it is not there.
#if DEBUG
struct ParallelSpaceBadge: View {
    var body: some View {
        Text(verbatim: "PARALLEL SPACE")
            .font(.system(size: 10, weight: .heavy))
            .tracking(1.5)
            .foregroundStyle(.black)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 3)
            .background(Color(red: 1, green: 0.84, blue: 0))
            .allowsHitTesting(false)
    }
}
#endif
